Feature/flash qmk

WildlyManifest opened the pull request:
I made these changes in order to flash my Atreus, which had QMK firmware and the Caterina bootloader that came with the AStar. 

The QMK firmware did not expose any serial ports until the keyboard was reset, so it could not be reset by sending a command from the script. A different method was necessary. 

The other changes were some improvements to the port detection logic for macos.

To flash an Atreus or another keyboard that has QMK, the QMK_FLASH flag needs to be set. In addition, the bootloader pid and vid needs to be set to the values defined by the bootloader that is currently installed. E.g. for my device:
```
QMK_FLASH=1
BOOTLOADER_VID="0x1ffb"
BOOTLOADER_PID="0x0101"
```

algernon commented on bin/kaleidoscope-builder:
```diff
@@ -160,37 +159,65 @@ prompt_before_flashing () {
+            # Keyboards with QMK firmware must be reset manually to enter bootloader mode.
+            echo "Reset your keyboard now to initiate flashing."
+            wait_for_bootloader_port
+        fi
```
So, the basic difference here is the manual reset, right? I'd suggest not tieing the name to QMK, because a manual reset can be useful even with Kaleidoscope. So how about we call it `MANUAL_RESET`? And narrow the paths a little, somewhat like his:

```suggestion
        prompt_before_flashing
        # This is defined in the (optional) user config.
        # shellcheck disable=SC2154
        ${preFlash_HOOKS}

        if [ -z "${MANUAL_RESET}" ]; then          
           reset_device
           sleep 2
           find_bootloader_ports
        else
           wait_for_bootloader_port
        fi
```

I do not think we need the extra text, because the pre-flashing prompt already tells us to put the keyboard into flash mode. A manual reset does that.

obra commented on bin/find-device-port-macos:
```diff
@@ -52,32 +54,33 @@ sub try_for_raw_serialnum {
-    # will need an update, as well.
-    my $loc = substr($location_id, 2, 3);
-    exit_with_port_if_exists("/dev/cu.usbmodem" . $loc . 1);
+    # will need an update, as well. 
```
I don't understand what this code change does or why. Can you add some more comments?

WildlyManifest replied:
From my limited experimentation, macOs does truncate the last character and replace it with '1' even if it is not a digit. This way, all possibilities are covered.

obra commented on bin/find-device-port-macos:
```diff
@@ -52,32 +54,33 @@ sub try_for_raw_serialnum {
-    exit_with_port_if_exists("/dev/cu.usbmodem" . $loc . 1);
+    # will need an update, as well. 
+    if ($location_id =~ /0x(\d+?)0*\b/) { 
+        exit_with_port_if_exists("/dev/cu.usbmodem" . $1 . 1);
```
using $1 directly there can be a little bit brittle and not assigning it to a variable misses an opportunity for a descriptive name to help make the code more readable.

What problem was this change fixing?

WildlyManifest replied:
The location string is not necessarily 3 chars long.

obra commented on bin/find-device-port-macos:
```diff
@@ -52,32 +54,33 @@ sub try_for_raw_serialnum {
 
 sub try_for_sn_prefix {
     my $sn = shift;
+    chop $sn if length $sn > 1; 
```
A bit of doc here, please

obra commented on bin/find-device-port-macos:
```diff
@@ -10,6 +10,8 @@ use strict;
 my $vid = shift @ARGV;
 my $pid = shift @ARGV;
 
+exit 0 unless defined $vid && defined $pid;
```
0 is a success condition for exiting. This should probably be something like 


```
if (!defined $vid || !defined $pid) {
  die "$0 has two required parameters, VID and PID";
}
```

obra commented:
Sorry it took a bit to review this.

We should clean up the history of this PR to have separate logical commits for each logical change. 

Whitespace changes, if any, should be in their own commits. 

For example, https://github.com/keyboardio/Kaleidoscope/pull/824/commits/1d1c18d65b186241a14ba7eb797fa967ca884c3b mixes a bunch of seemingly unrelated changes and whitespace changes.

I don't understand the reasons behind some of the changes to the OSX port detection logic. If there are specific issues being solved, ideally doc them inline. Failing that, explain them in the commit message.

Thanks!

WildlyManifest commented:
Hi Jesse, 
I think I have addressed your review comments. The only thing I am not sure about is cleaning up the commits; I dont really know how to go about doing that.

Jason Koh

obra commented:
Hi Jason,

Thanks so much for the updates!
Don't worry about that. I'm happy to take care of it :)

For the future, "git rebase -i origin/master" is the command that kicks off
that kind of workflow -
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History is a tutorial
that might help some, but this is very much one of the bits of git
that's....well, I don't want to say that it's not user friendly, but it's
much pickier about who its friends are.

Best,
Jesse

On Sun, May 3, 2020 at 2:10 AM Jason Koh <notifications@github.com> wrote:

> Hi Jesse,
> I think I have addressed your review comments. The only thing I am not
> sure about is cleaning up the commits; I dont really know how to go about
> doing that.
>
> Jason Koh
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/824#issuecomment-623079222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2EDBPLMOGWBLT7GFGTRPUYJNANCNFSM4LQ4XH3A>
> .
>

obra commented:
@WildlyManifest Wanna have a look at #841 - I took your branch, redid the history and cleaned up a couple more things. I should be able to test it on a real mac later today. I'm closing this PR, so discussion should move over there.  Thanks so much for all the hacking!
